### ISSUE TITLE
using rules_scala documented scalafmt support breaks scala dependency discovery

### ISSUE THREAD

**Issue by gergelyfabian** [bazelbuild/intellij#1824]
If you define a scalafmt rule as documented in rules_scala it won't be recognized by intellij plugin as a Scala target, as the rule names for rules_scala is hard-coded.

https://github.com/bazelbuild/rules_scala/blob/master/docs/phase_scalafmt.md

E.g. you define and use scalafmt_scala_binary or scalafmt_scala_library rules, then those won't be recognized as Scala, but rather Java rules.

In effect the external dependencies for such targets are not discovered by IntelliJ.

The reason is that the Scala targets generated by rules_scala do not expose a jdeps interface, hence intellij BlazeJavaWorkspaceImporter cannot find the dependencies for them (assuming they are not in the working set).

I've created a possible fix in https://github.com/bazelbuild/intellij/compare/master...gergelyfabian:fix-rules_scala-scalafmt-rule-detection#diff-08b43ae09d0d1e0fe6151b8610450a37.

**Comment by gergelyfabian** [comment]
As rules_scala maintainer, @ittaiz, what do you think about this? Should it be solved this way, or do we need some changes also to rules_scala, like making the scalafmt functionality into officially supported rule names?
Or maybe the documentation should be changed to say that I should create rule names that I know that the plugin would recognize:

```
load("//scala:advanced_usage/scala.bzl", "make_scala_binary", "make_scala_library")
load("//scala/scalafmt:phase_scalafmt_ext.bzl", "ext_scalafmt")

scala_binary = make_scala_binary(ext_scalafmt)
scala_library = make_scala_library(ext_scalafmt)
```

And then use these alternative scala_library/scala_binary rules...

**Comment by ittaiz** [comment]
That’s a great question.
Q- if we would have exposed jdeps then it wouldn’t have mattered how the rule is called (assuming JavaInfo of course)?
If so I think that maybe we should consider exposing jdeps (optionally maybe?).

Main problem is that the IJ team capacity for community related changes like generic support of scala or something is very limited these days (AFAIK)

**Comment by gergelyfabian** [comment]
Yes, I think jdeps support would solve the issue. It works ok for Java, and
Scala rule names are detected as Java by default (if the name happens to be
not hard coded in the intellij plugin's Scala rule detection).

My example fix any how would easily break if someone would use the
rules_scala rule macros in another way, and call the created rules in
another way (it fixes the issue for now though).

Also, the potential jdeps support in rules_scala would mean, that the
intellij plugin wouldn't need to have this hard coded logic for Scala rule
name detection, and maybe it could even be treated as general Java targets?
In any case it seems to remove complexity in the big picture instead of
adding it, moving to a common API.

It would be great to have this kind of support in rules_scala..

Ittai Zeidman <notifications@github.com> ezt írta (időpont: 2020. máj. 15.,
P 19:55):

**Comment by ittaiz** [comment]
1. I think custom scala support was needed for a few reasons and not only target identification so maybe more will be lost?
2. Need to remember jdeps support doesn’t come for free (impl, maintenance, performance)
3. Sounds like a documentation PR in rules_scala would definitely be valuable for now

**Comment by gergelyfabian** [comment]
Regarding the documentation PR, what should be the general advice at this moment?

Mention that intellij plugin supports only a set of names, and using custom-named rules, defined by using macros (including the call to add scalafmt support) makes intellij plugin not recognize those as Scala targets? That if you need extended support from rules_scala (including scalafmt), you should rather name those rules the same way as in rules_scala (and use them from your own scope)? Is this the best workaround we have currently? Anything else?

**Comment by ittaiz** [comment]
I think so.
Anything else you can think of?

**Comment by gergelyfabian** [comment]
I sent the documentation PR: https://github.com/bazelbuild/rules_scala/pull/1047

**Comment by odisseus** [comment]
I found a workaround for this issue that does not require forking the entire Bazel plugin.

It is possible to create a tiny plugin that would [extend](https://plugins.jetbrains.com/docs/intellij/plugin-extensions.html) the Bazel plugin with a custom [TargetKindProvider](https://github.com/bazelbuild/intellij/blob/v2022.05.03/base/src/com/google/idea/blaze/base/model/primitives/Kind.java#L53) which would define kinds for custom rule names.

This should go into `plugin.xml`:
```xml
<extensions defaultExtensionNs="com.google.idea.blaze">
  <TargetKindProvider implementation="com.foobar.bazel.kind.FoobarScalaBlazeRules"/>
</extensions>
```

The kind provider code could look like this:
```scala
package com.foobar.bazel.kind;

import com.google.common.collect.ImmutableSet;
import com.google.idea.blaze.base.model.primitives.Kind;
import com.google.idea.blaze.base.model.primitives.LanguageClass;
import com.google.idea.blaze.base.model.primitives.RuleType;

public class FoobarScalaBlazeRules implements Kind.Provider {

  private static final Kind SCALA_FOOBAR_LIBRARY =
      Kind.Provider.create("_scala_foobar_library", LanguageClass.SCALA, RuleType.LIBRARY);

  @Override
  public ImmutableSet<Kind> getTargetKinds() {
    return ImmutableSet.of(SCALA_FOOBAR_LIBRARY);
  }
}
```
Of course, you would have to build this plugin and distribute it within your organization, but this is probably easier than maintaining a fork of the Bazel plugin and tracking the upstream changes.

**Comment by gergelyfabian** [comment]
I think this is still valid.
